packages/model-viewer/src/three-components/ARRenderer.ts

```diff
@@ -248,117 +327,330 @@ export class ARRenderer extends EventDispatcher {
+    // position is not updated when matrix is updated.
+    camera.position.setFromMatrixPosition(cameraMatrix);
+
+    if (this[$initialHitSource] != null) {
+      // Target locked to screen center
+      const {position, model} = this[$presentedScene]!;
+      const radius = model.idealCameraDistance;
+      camera.getWorldDirection(position);
+      position.multiplyScalar(radius);
+      position.add(camera.position);
     }
-    // NOTE: Currently rays will be cast from the middle of the screen.
-    // Eventually we might use input coordinates for this.
 
-    // Just reuse the hit matrix that the reticle has computed.
-    if (this.reticle && this.reticle.hitMatrix) {
+    if (!this[$initialized]) {
+      camera.projectionMatrix.fromArray(view.projectionMatrix);
+      // Have to set the inverse manually when setting matrix directly. This is
+      // needed for raycasting.
+      camera.projectionMatrixInverse.getInverse(camera.projectionMatrix);
+      // Orient model toward camera on first frame.
+      const {x, z} = camera.position;
       const scene = this[$presentedScene]!;
-      const {model} = scene;
-      const {hitMatrix} = this.reticle;
+      scene.pointTowards(x, z);
+      scene.model.updateMatrixWorld(true);
+      this[$goalYaw] = scene.yaw;
+      this[$initialModel2World].copy(scene.model.matrixWorld);
+      this[$initialized] = true;
+    }
 
-      scene.position.setFromMatrixPosition(hitMatrix);
+    this[$presentedScene]!.model.orientHotspots(
+        Math.atan2(cameraMatrix.elements[1], cameraMatrix.elements[5]));
+  }
 
-      // Orient the scene to face the camera
-      const camPosition = vector3.setFromMatrixPosition(this.camera.matrix);
-      scene.pointTowards(camPosition.x, camPosition.z);
-      scene.updateMatrixWorld();
+  [$placeInitially](frame: XRFrame) {
+    const hitSource = this[$initialHitSource];
+    if (hitSource == null) {
+      return;
+    }
 
-      model.visible = true;
-      model.setHotspotsVisibility(true);
+    const hitTestResults = frame.getHitTestResults(hitSource);
+    if (hitTestResults.length == 0) {
+      return;
+    }
 
-      this.dispatchEvent({type: 'modelmove'});
+    const hit = hitTestResults[0];
+    const hitMatrix = this[$getHitPoint](hit);
+    if (hitMatrix == null) {
+      return;
     }
+
+    this.placeModel(hitMatrix);
+
+    hitSource.cancel();
+    this[$initialHitSource] = null;
+
+    const {session} = frame;
+    session.addEventListener('selectstart', this[$selectStartHandler]);
+    session.addEventListener('selectend', this[$selectEndHandler]);
+    session
+        .requestHitTestSourceForTransientInput({profile: 'generic-touchscreen'})
+        .then(hitTestSource => {
+          this[$transientHitTestSource] = hitTestSource;
+        });
+  }
+
+  [$getHitPoint](hitResult: XRHitTestResult): Vector3|null {
+    const hitMatrix = matrix4.fromArray(
+        hitResult.getPose(this[$refSpace]!)!.transform.matrix);
```

Comment:
That's a lot of possible `null` that we aren't guarding against. I think we should at least check that `getPose` gave us something...

Reply:
SG; I'll blame that on inherited code, I think...